client/src/procedures.rs

```diff
@@ -0,0 +1,440 @@
+}
+
+// ==========================
+//  Traits for the `Procedure` derive-macro
+// ==========================
+
+pub struct Products<T> {
+    pub secret: Vec<u8>,
+    pub output: T,
+}
+
+trait Parse {
+    type Input;
+    type Output;
+    fn parse(self, input: Self::Input) -> Result<Self::Output, engine::Error>;
+}
+
+trait Generate {
+    type Input;
+    type Output;
+    fn generate(self, input: Self::Input) -> Result<Products<Self::Output>, engine::Error>;
+}
+
+trait Process {
+    type Input;
+    type Output;
+    fn process(self, input: Self::Input, guard: GuardedVec<u8>) -> Result<Products<Self::Output>, engine::Error>;
+}
+
+trait Utilize {
+    type Input;
+    type Output;
+    fn utilize(self, input: Self::Input, guard: GuardedVec<u8>) -> Result<Self::Output, engine::Error>;
+}
+
+// ==========================
+//  Input /  Output Info
+// ==========================
+
+#[derive(Clone, Serialize, Deserialize)]
+pub enum InputData<T>
+where
+    T: TryFrom<ProcedureIo>,
+{
+    Key(OutputKey),
+    Value(T),
+}
+
+#[derive(Clone, Serialize, Deserialize)]
+pub struct InterimProduct<T> {
+    pub target: T,
+    pub is_temp: bool,
+}
+
+pub trait SourceInfo {
+    fn source_location(&self) -> &Location;
+    fn source_location_mut(&mut self) -> &mut Location;
+}
+
+impl SourceInfo for Location {
+    fn source_location(&self) -> &Location {
+        self
+    }
+    fn source_location_mut(&mut self) -> &mut Location {
+        self
+    }
+}
+
+impl SourceInfo for SLIP10DeriveInput {
+    fn source_location(&self) -> &Location {
+        match self {
+            SLIP10DeriveInput::Seed(l) => l,
+            SLIP10DeriveInput::Key(l) => l,
+        }
+    }
+
+    fn source_location_mut(&mut self) -> &mut Location {
+        match self {
+            SLIP10DeriveInput::Seed(l) => l,
+            SLIP10DeriveInput::Key(l) => l,
+        }
+    }
+}
+
+pub trait TargetInfo {
+    fn target_info(&self) -> &InterimProduct<Target>;
+    fn target_info_mut(&mut self) -> &mut InterimProduct<Target>;
+
+    fn target(&self) -> Location {
+        self.target_info().target.location.clone()
+    }
+
+    fn write_secret(mut self, location: Location, hint: RecordHint) -> Self
+    where
+        Self: Sized,
+    {
+        let target = self.target_info_mut();
+        target.target = Target { location, hint };
+        target.is_temp = false;
+        self
+    }
+}
+
+#[derive(Clone, Serialize, Deserialize)]
+pub struct Target {
+    pub location: Location,
+    pub hint: RecordHint,
+}
+
+impl Target {
+    pub fn random() -> Self {
+        let location = Location::generic(Self::random_path(), Self::random_path());
+        let hint = RecordHint::new("".to_string()).unwrap();
+        Target { location, hint }
+    }
+
+    fn random_path() -> impl Into<Vec<u8>> {
+        let l = (rand::random::<u8>() % (u8::MAX - 1)) + 1;
+        let mut v = Vec::with_capacity(l.into());
+        for _ in 0..l + 1 {
```

Comment:
the increment is not needed here, as you already make sure that `l > 0` will always be true.